scripts/test/Muon/threading_manager_test.py

```diff
@@ -0,0 +1,303 @@
+
+    def test_that_cannot_split_empty_list(self):
+        lst = []
+        split_list = split_list_into_n_parts(lst, 4)
+        self.assertEqual(len(split_list), 4)
+        for sublist in split_list:
+            self.assertEqual(sublist, [])
+
+    def test_split_kwarg_list_with_list_of_several_kwargs(self):
+        kwarg_list = {"arg1": [1, 2, 3, 4, 5], "arg2": ["a", "b", "c", "d", "e"]}
+        split_kwargs = split_kwarg_list(kwarg_list, 2)
+
+        self.assertEqual(len(split_kwargs), 2)
+        self.assertEqual(split_kwargs[0], {"arg1": [1, 3, 5], "arg2": ["a", "c", "e"]})
+        self.assertEqual(split_kwargs[1], {"arg1": [2, 4], "arg2": ["b", "d"]})
+
+
+class ThreadingManagerWorkerTest(unittest.TestCase):
+    class Runner:
+        QT_APP = QtWidgets.QApplication([])
+
+        def __init__(self, thread):
+            self._thread = thread
+            self._thread.signals.finished.connect(self.finished)
+            self.QT_APP.exec_()
+
+        def finished(self):
+            self.QT_APP.processEvents()
+            self.QT_APP.exit(0)
+
+    def setUp(self):
+        pass
+
+    def test_that_worker_thread_starts_and_finishes(self):
+        worker = Worker(test_function, lst=[1, 2, 3, 4, 5])
+
+        worker_started = mock.Mock()
+        worker_finished = mock.Mock()
+
+        worker.signals.started.connect(worker_started)
+        worker.signals.finished.connect(worker_finished)
+        worker.start()
+        self.Runner(worker)
+
+        self.assertEqual(worker_started.call_count, 1)
+        self.assertEqual(worker_finished.call_count, 1)
+
+    def test_that_worker_thread_updates_progress_from_each_evaluation(self):
+        worker = Worker(test_function, lst=[1, 2, 3, 4, 5])
+
+        worker_progress = mock.Mock()
+
+        worker.signals.progress.connect(worker_progress)
+
+        worker.start()
+        self.Runner(worker)
+
+        self.assertEqual(worker_progress.call_count, 5)
+
+    def test_that_worker_thread_evaluates_results_correctly(self):
+        worker = Worker(test_function, lst=[1, 2, 3, 4, 5])
+
+        worker_result = mock.Mock()
+
+        worker.signals.result.connect(worker_result)
+
+        worker.start()
+        self.Runner(worker)
+
+        self.assertEqual(worker_result.call_count, 1)
+        self.assertEqual(worker_result.call_args_list[0][0][0][0]["lst"], [1, 2, 3, 4, 5])
+        self.assertEqual(worker_result.call_args_list[0][0][0][1], [2, 4, 6, 8, 10])
+
+    def test_that_error_slot_called_if_function_raises(self):
+        worker = Worker(test_function_raises_on_n, lst=[1, 2, 3, 4, 5], raise_on=1)
+
+        worker_result = mock.Mock()
+        worker_error = mock.Mock()
+
+        worker.signals.result.connect(worker_result)
+        worker.signals.error.connect(worker_error)
+
+        worker.start()
+        self.Runner(worker)
+
+        self.assertEqual(worker_result.call_count, 0)
+        self.assertEqual(worker_error.call_count, 1)
```

Comment:
`worker_result`, `worker_error`, `worker_progress`, `worker_started` and `worker_finished` are used/shared throughout the tests. I'd move these to `setUp()` and make them elements of `self` to clear up the test functions.

Consider creating a function that accepts `Worker(...)` and connects would-be `self.worker.signals...` , calls `worker.start()` and calls `self.Runner(worker)` as these are common to all tests. This could then be called before the asserts.

Reply:
All signals could be connected in such a function, and only the outputs required would be checked in the `self.assertEquals`